### PR TITLE
Make system module map generation faster and fully hermetic

### PR DESCRIPTION
**Pull request by fmeum**
* `system_module_maps` no longer performs any IO.
* The generated module map no longer references any non-hermetic paths and can thus be cached remotely, even when toolchain or sysroot are provided as absolute paths.

### CODE REVIEW

File: toolchain/toolchains.bzl.tpl [hunk 1]
```diff
@@ -12,7 +12,13 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+load("@bazel_features//:deps.bzl", "bazel_features_deps")
+
 def llvm_register_toolchains():
+    # This doesn't really belong here, but it's the only place to make this
```
Comment (fmeum) [inline toolchain/toolchains.bzl.tpl:18]: This is a hack for the sake of backwards compatibility. If you see a better way or, alternatively, would consider adding a new macro call to the WORKSPACE setup, please let me know.

File: toolchain/internal/system_module_map.bzl [hunk 2]
```diff
@@ -12,64 +12,97 @@
...
+        )
+        template_dict.add_joined(
+            "%umbrella_submodules%",
+            depset(absolute_path_dirs),
+            join_with = "\n",
+            map_each = _umbrella_submodule,
+        )
+        expand_template_kwargs = {"computed_substitutions": template_dict}
+    else:
+        # Memory-inefficient, but works on older Bazel versions that only
```
Comment (siddharthab) [inline toolchain/internal/system_module_map.bzl:74]: We have now moved the minimum supported Bazel version of this repo to be 7.0.0.

File: toolchain/toolchains.bzl.tpl (hunk 1 above)
Comment (siddharthab) [reply toolchain/toolchains.bzl.tpl:18]: Can this not go in toolchain/deps.bzl?

**Review by siddharthab — APPROVED** [review]
Please land whenever you are ready. Let's make a release when this lands.

File: toolchain/internal/system_module_map.bzl (hunk 2 above)
Comment (fmeum) [reply toolchain/internal/system_module_map.bzl]: Thanks, I can just get rid of the troublesome dep then.

File: toolchain/internal/configure.bzl [hunk 3]
```diff
@@ -586,3 +593,9 @@ native_binary(
         llvm_dist_label_prefix = llvm_dist_label_prefix,
         host_dl_ext = host_dl_ext,
     )
+
+def _is_hermetic_or_exists(rctx, path, sysroot_prefix):
```
[the remark this answers is not in the document]
Comment (fmeum) [reply toolchain/internal/configure.bzl:597]: I didn't have this scenario in mind when I wrote this, my assumption being that you would only use this toolchain for RBE together with a hermetic sysroot. 

This function only makes sense if you are actually targetting the host, so we should probably disable it when cross compiling and just use all directories unfiltered. If you want to give that a try, I can help you implement and test this approach.

[the remark this answers is not in the document]
Comment (honnix) [reply toolchain/internal/configure.bzl:597]: I see. I'm not sure of a reasonable implementation. What would you suggest?

[the remark this answers is not in the document]
Comment (fmeum) [reply toolchain/internal/configure.bzl:597]: The `_is_hermetic_or_exists` function should be fed the information whether [`exec_os`](https://github.com/bazel-contrib/toolchains_llvm/blob/e226bfa2800f1705b68bb0eb238326edf6a3ec57/toolchain/internal/configure.bzl#L525) equals [`target_os`](https://github.com/bazel-contrib/toolchains_llvm/blob/e226bfa2800f1705b68bb0eb238326edf6a3ec57/toolchain/internal/configure.bzl#L523) and only perform the existence check if so.

[the remark this answers is not in the document]
Comment (honnix) [reply toolchain/internal/configure.bzl:597]: Yeah I was looking at those, but not sure that is reliable. For RBE, `exec_os` could be the same as `target_os`, right? For example I can define a toolchain like the following and while produce target for `linux` as well.

```bazel
llvm.toolchain(
    name = "llvm_toolchain_linux_exec",
    exec_arch = "amd64",
    exec_os = "linux",
    llvm_version = LLVM_VERSION,
)
```

[the remark this answers is not in the document]
Comment (fmeum) [reply toolchain/internal/configure.bzl:597]: You are completely right, I guess we need to match the host OS (obtained e.g. from `repository_ctx.os`) against the target OS (this determines the shape of the sysroot).

[the remark this answers is not in the document]
Comment (honnix) [reply toolchain/internal/configure.bzl:597]: Let me give that a try.

[the remark this answers is not in the document]
Comment (honnix) [reply toolchain/internal/configure.bzl:597]: `rctx.os.name` for mac is like `mac os x`, which is not the same as the convention used by `target_os` :(

**Pull request merged**